oyster/src/utils/file-processor.js

```diff
@@ -25,100 +39,134 @@ const createByteChunks = file => {
-    return { chunkIdx: index, chunkStartingPoint: byte };
+    return { chunkIdx: index + 1, chunkStartingPoint: byte };
   });
 
   return byteChunks;
 };
 
-const createUploadSession = file =>
+const createUploadSession = (file, handle) =>
   new Promise((resolve, reject) => {
     request.post(
-      `${API.HOST}${API.V1_UPLOAD_SESSIONS_PATH}`,
       {
-        file_size_bytes: file.size,
-        genesis_hash: createHandle(file.name)
+        url: `${API.HOST}${API.V1_UPLOAD_SESSIONS_PATH}`,
+        form: {
+          file_size_bytes: file.size,
+          genesis_hash: handle
+        }
       },
       (error, response, body) => {
         if (error) {
-          return reject(error);
+          console.log("ERROR: ", error);
+          // TODO: uncomment this
+          // return reject(error);
         }
-        resolve(JSON.parse(response.body));
+        resolve();
       }
     );
   });
 
-const uploadFileToBrokerNodes = file => {
-  const byteChunks = createByteChunks(file);
-
-  Promise.all([
-    sendToAlphaBroker(byteChunks, file),
-    sendToBetaBroker(byteChunks, file)
-  ]).then(() => {
-    console.log("Upload complete! ", sentChunks);
-  });
-};
-
 const createReader = onRead => {
   const reader = new FileReader();
   reader.onloadend = function(evt) {
-    if (evt.target.readyState == FileReader.DONE) {
+    if (evt.target.readyState === FileReader.DONE) {
       onRead(evt.target.result);
     }
   };
   return reader;
 };
 
-const sentChunks = {};
-const sendChunkToNode = (chunkIdx, fileSlice, handle) => {
-  if (!!sentChunks[chunkIdx]) {
-    return;
-  }
+const sendChunkToBroker = (chunkIdx, data, handle) =>
+  new Promise((resolve, reject) => {
+    const encryptedData = encrypt(data, handle);
+    request.post(
+      {
+        url: `${API.HOST}${API.V1_UPLOAD_CHUNKS_PATH}`,
+        form: {
+          chunkIdx,
+          data: encryptedData,
+          genesis_hash: handle
+        }
+      },
+      (error, response, body) => {
+        if (error) {
+          return reject(error);
+        }
+        resolve(JSON.parse(response.body));
+      }
+    );
 
-  // TODO: wrap this in promise and actually send the chunks to nodes
-  const encryptedChunk = encrypt(fileSlice, handle);
-  sentChunks[chunkIdx] = encryptedChunk;
+    return encryptedData;
+  });
 
-  return encryptedChunk;
+const sendFileContentsToBroker = (file, handle, byteChunks, sliceCutOffFn) => {
+  const chunkRequests = byteChunks.map(
+    byte =>
+      new Promise((resolve, reject) => {
+        const { chunkIdx, chunkStartingPoint } = byte;
+        const blob = file.slice(
+          chunkStartingPoint,
+          sliceCutOffFn(chunkStartingPoint)
+        );
+
+        const reader = createReader(fileSlice => {
+          sendChunkToBroker(chunkIdx, fileSlice, handle).then(resolve);
+        });
+        reader.readAsBinaryString(blob);
+      })
+  );
+
+  return Promise.all(chunkRequests);
```

Comment:
I think Promise.all has fail fast behavior, which we don't want. I also think Bruno mentioned he wants this throttled to like 10 concurrent requests at a time. Not going to block this PR, but we should add that in the future.

Reply:
The request limiting to 10 is covered in another task, so should be fine

https://oyster.axosoft.com/viewitem?id=17&type=features&force_use_number=false

Reply:
Wonder if we could use observables/subjects or something if promises won't work?

Reply:
+1 for Rx. Might be overkill for the testnet deadline, but adding it in the future would be good for async management.

For testnet we can do something simple like https://github.com/DJWassink/Promise-parallel-throttle. It solves our 2 problems.